USB DFU permanent download and automatic reboot

hackwerken opened the pull request:
Proposed solution for  #41921.

Also updated some outdated information in the readme of the USB DFU sample.

hackwerken committed:
usb: dfu: add permanent download and automatic reboot

This commit adds the USB_DFU_PERMANENT_DOWNLOAD and USB_DFU_REBOOT and
symbols.When the permanent download symbol is enabled, slot 1 will be
marked as confirmed. With the reboot symbol enabled, the devices
automatically reboots after the download is completed.

The functionality is split into two symbols to allow the automatic
reboot without confirming the image. This enables image confirmation via
another channel. For example via the shell’s Mcuboot commands.

This functionality allows downloading an image to the device without the
user having to interact with the device. It is useful in cases where
ease of use is more important then safety. For example  when using USB
download for daily development. This is especially applicable for
devices with a closed case.

The changes were tested on an nrf52840dk. The following line can be used
to build the USB DFU example with the symbols enabled.

west build -b nrf52840dk_nrf52840 zephyr/samples/subsys/usb/dfu \
-d build-dfu -- -DCONFIG_BOOTLOADER_MCUBOOT=y \
-DOVERLAY_CONFIG=overlay-reboot-permanent.conf \
-DCONFIG_MCUBOOT_SIGNATURE_KEY_FILE\
=\"bootloader/mcuboot/root-rsa-2048.pem\"

Fixes #41921

Signed-off-by: Martijn Stommels <martijn@martijnpc.nl>

hackwerken committed:
usb: dfu: update sample README

Some information in the USB DFU sample was out of date.

Also added some example lines for the Permanent download and automatic
reboot features.

Signed-off-by: Martijn Stommels <martijn@martijnpc.nl>

jfischer-no commented:
>`Signed-off-by: Martijn Stommels <1781040+hackwerken@users.noreply.github.com>`

@hackwerken AFAIK github `username` from `users.noreply` subdomain is not acceptable in Zephyr Project.

hackwerken commented:
@jfischer-no according to the CI checks you are right :).

I also fixed the other issues it found.

hackwerken commented:
@jfischer-no thanks for running the workflows.

I removed the newline at the end of the sample readme.

I'm not sure why the footprint delta was failing, since it was warning about Kconfig symbols I didn't toch... I rebased my branch, so hopefully that helps.

hackwerken commented:
@nvlsianpu thanks for your review. I implemented your suggestions. 

I also ran the check concerning the newline locally. It should be fixed now.

jfischer-no commented on subsys/usb/class/dfu/usb_dfu.c:
```diff
@@ -374,7 +389,12 @@ static void dfu_flash_write(uint8_t *data, size_t len)
 		dfu_reset_counters();
-		if (boot_request_upgrade(false)) {
+
+		const bool should_confirm = IS_ENABLED(CONFIG_USB_DFU_PERMANENT_DOWNLOAD);
```
Please move variable definition to top of the block
```diff c
diff --git a/subsys/usb/class/dfu/usb_dfu.c b/subsys/usb/class/dfu/usb_dfu.c
index 42ead13b49e..4cc3eeea56c 100644
--- a/subsys/usb/class/dfu/usb_dfu.c
+++ b/subsys/usb/class/dfu/usb_dfu.c
@@ -382,14 +382,13 @@ static void dfu_flash_write(uint8_t *data, size_t len)
                dfu_data.state = dfuERROR;
                dfu_data.status = errWRITE;
        } else if (!len) {
+               const bool should_confirm = IS_ENABLED(CONFIG_USB_DFU_PERMANENT_DOWNLOAD);
+
                LOG_DBG("flash write done");
                dfu_data.state = dfuMANIFEST_SYNC;
                dfu_reset_counters();
 
-               const bool should_confirm = IS_ENABLED(CONFIG_USB_DFU_PERMANENT_DOWNLOAD);
-
                LOG_DBG("Should confirm: %d", should_confirm);
-
                if (boot_request_upgrade(should_confirm)) {
                        dfu_data.state = dfuERROR;
                        dfu_data.status = errWRITE;
```

jfischer-no commented on subsys/usb/class/dfu/usb_dfu.c:
```diff
@@ -407,7 +427,20 @@ static int dfu_class_handle_to_host(struct usb_setup_packet *setup,
 			dfu_data.status, dfu_data.state);
 
 		if (dfu_data.state == dfuMANIFEST_SYNC) {
+
+#if IS_ENABLED(CONFIG_USB_DFU_REBOOT)
+			dfu_data.state = dfuMANIFEST_WAIT_RST;
+
+			LOG_DBG("Scheduling reboot in 500ms");
+
+			/*
+			 * Reboot with a delay,
+			 * so there is some time to send the status to the host
+			 */
+			k_work_schedule_for_queue(&USB_WORK_Q, &reboot_work, K_MSEC(500));
+#else
```
Maybe it will be a bit clearer that way:

```diff
diff --git a/subsys/usb/class/dfu/usb_dfu.c b/subsys/usb/class/dfu/usb_dfu.c
index 1fe835cd020..42ead13b49e 100644
--- a/subsys/usb/class/dfu/usb_dfu.c
+++ b/subsys/usb/class/dfu/usb_dfu.c
@@ -99,10 +99,6 @@ struct dfu_worker_data_t {
 
 static struct dfu_worker_data_t dfu_data_worker;
 
-#ifdef CONFIG_USB_DFU_REBOOT
-static struct k_work_delayable reboot_work;
-#endif
-
 struct usb_dfu_config {
        struct usb_if_descriptor if0;
        struct dfu_runtime_descriptor dfu_descr;
@@ -414,6 +410,30 @@ static void dfu_timer_expired(struct k_timer *timer)
        }
 }
 
+#ifdef CONFIG_USB_DFU_REBOOT
+static struct k_work_delayable reboot_work;
+
+static void reboot_work_handler(struct k_work *item)
+{
+       ARG_UNUSED(item);
+
+       sys_reboot(SYS_REBOOT_WARM);
+}
+
+static void reboot_schedule(void)
+{
+       dfu_data.state = dfuMANIFEST_WAIT_RST;
+
+       LOG_DBG("Scheduling reboot in 500ms");
+
+       /*
+        * Reboot with a delay,
+        * so there is some time to send the status to the host
+        */
+       k_work_schedule_for_queue(&USB_WORK_Q, &reboot_work, K_MSEC(500));
+}
+#endif
+
 static int dfu_class_handle_to_host(struct usb_setup_packet *setup,
                                    int32_t *data_len, uint8_t **data)
 {
@@ -427,17 +447,8 @@ static int dfu_class_handle_to_host(struct usb_setup_packet *setup,
                        dfu_data.status, dfu_data.state);
 
                if (dfu_data.state == dfuMANIFEST_SYNC) {
-
 #if IS_ENABLED(CONFIG_USB_DFU_REBOOT)
-                       dfu_data.state = dfuMANIFEST_WAIT_RST;
-
-                       LOG_DBG("Scheduling reboot in 500ms");
-
-                       /*
-                        * Reboot with a delay,
-                        * so there is some time to send the status to the host
-                        */
-                       k_work_schedule_for_queue(&USB_WORK_Q, &reboot_work, K_MSEC(500));
+                       reboot_schedule();
 #else
                        dfu_data.state = dfuIDLE;
 #endif
@@ -861,15 +872,6 @@ static void dfu_work_handler(struct k_work *item)
        }
 }
 
-#ifdef CONFIG_USB_DFU_REBOOT
-static void reboot_work_handler(struct k_work *item)
-{
-       ARG_UNUSED(item);
-
-       sys_reboot(SYS_REBOOT_WARM);
-}
-#endif
-
 static int usb_dfu_init(const struct device *dev)
 {
        const struct flash_area *fa;
```

hackwerken commented:
@jfischer-no I implemented your feedback. Having all the reboot stuff together is much nicer indeed.

I left the state change outside the reboot_schedule() function, so the state machine is handled all at the same place. This also means the functions doesn't have side effects that aren't apparent from it's name.

hackwerken commented:
@nvlsianpu thanks for the CI run.

I fixed the comment alignment. Also, apparently `:kconfig` changed to `:kconfig:option:` for the documentation, so fixed that as well.

jfischer-no commented on subsys/usb/class/dfu/usb_dfu.c:
```diff
@@ -53,6 +53,10 @@
 #include <usb_work_q.h>
 
 #define LOG_LEVEL CONFIG_USB_DEVICE_LOG_LEVEL
+#ifdef CONFIG_USB_DFU_REBOOT
```
No need to include `#include <sys/reboot.h>` conditional.

hackwerken replied:
Fixed.
